### ISSUE TITLE
Investigate running tests in parallel with parallel mesh

### ISSUE THREAD

**Issue by permcody** [idaholab/moose#1500]
We need to look at which parts of the system do not work properly with parallel mesh in parallel.

So far it appears that there are issues with the following systems:
- geometric search
- periodic boundaries (libMesh - very low priority)
- constant monomials (???)

**Comment by roystgnr** [comment]
See, this is why I need to be Cc:ed on things.

**Comment by permcody** [comment]
This is coming! It's on my TODO list when I get back along with getting "moar better hardware"!

**Comment by YaqiWang** [comment]
@roystgnr Is it possible to make `BreakBoundaryOnSubdomain` mesh modifier work? Several of my tests relying on that. Looks to me we should be able to make it work. I just do not understand how boundary info should be constructed in distributed mesh. Is boundary info different on processors?

**Comment by YaqiWang** [comment]
Another question is that can mesh-only option coexist with distributed mesh? Or will the distributed runs generate exodus output properly?

**Comment by permcody** [comment]
`--mesh-only` should work with DistributedMesh. The output format and "run time" objects are independent.

**Comment by YaqiWang** [comment]
How is the `allow_renumbering=true` in `MooseMesh` working? I want to understand why turning it to false with distributed mesh gives me the correct results while the default value does not. There is no obvious reason to me why this parameter can affect the calculation.

**Comment by permcody** [comment]
What do you mean by correct result? As far as your physics go, there should be little difference to you whether nodes are renumbered or not. If however you are trying to use node numbering, things will definitely change.

This is a complicated subject with lots of history. We have very convoluted and possibly incorrect logic in MOOSE mesh to adjust that value depending on several factors. You can look at the logic yourself. We may need to adjust it even more moving forward as it seems there are an increasing number of people with issues.

**Comment by YaqiWang** [comment]
Oh, wait, I might know why I need to turn allow_renumbering to false. I am assigning block IDs based on element IDs. Nothing wrong with other moose objects.

**Comment by roystgnr** [comment]
That sounds like a likely culprit.  While debugging I've run into several MOOSE objects that (at least as a popular option, if not as the only option) constructed problem based on input files listing specific element IDs.  The other typical reason to disallow renumbering is for output files - in particular the CSV output appears to be ordered by node id, and thus becomes incomprehensible if you both allow renumbering and use options like DistributedMesh which make renumbering more likely.

**Comment by roystgnr** [comment]
Oh, wait, I missed an earlier question:

When attached to a DistributedMesh, BoundaryInfo objects do indeed only store the boundary info data which is relevant on the elements and nodes that are semilocal to (i.e. owned by or with ghosted copies on) the current processor.  Usually this is fine (you can't query for boundary info using a pointer to an element you don't have, for instance), but it does mean that many methods behave differently depending on whether your mesh is currently distributed: n_boundary_ids(), build_*_boundary_ids(), build_*_list(), get_*boundary_ids().  A few methods give answers about the global boundary info directly: n_*_conds().  I think boundary_info.h now correctly documents every method's behavior, so when you aren't sure whether a method will give you a semilocal or a global result you can check there.

**Comment by roystgnr** [comment]
Damn I need to be more careful about how I write my commit and PR messages.

**Comment by roystgnr** [comment]
This is probably ready to go (partly) into CI.  We should start high-processor-count testing with ReplicatedMesh, as per #9430, but last I checked, the test directory (except the tests that are specifically set to replicated) now passes with --distributed-mesh too, and we should get Civet to ensure that doesn't regress.

I haven't tested --distributed-mesh with the modules directory yet, so this issue should remain open until that's done.

**Comment by roystgnr** [comment]
@brianmoose, could you do the same manual "sweep from 1 processor to 16" run that you did for #9430, but with --distributed-mesh turned on?

I suspect that what we'll want in the long run is "-np 4 --distributed-mesh" instead of the current "-np 2" on PRs, and then a nightly or a weekly sweep in both replicated and distributed forms.

**Comment by brianmoose** [comment]
No problem. I started it up here : https://www.moosebuild.org/job/104528/

**Comment by roystgnr** [comment]
Thanks!

**Comment by roystgnr** [comment]
Huh.  The tests are running slightly *faster* with --distributed-mesh?  I would have thought they would be *way* too small for the better scalability to outweigh the greater overhead.

Looks like we've got a couple dozen tests set to "skipped (MESH_MODE!=REPLICATED)"?  That may be an explanation for the speed difference, and that is definitely a bug I need to fix ...or two bugs I need to fix, if there are cases which actually still don't work in a distributed mode.

**Comment by roystgnr** [comment]
There are skipped cases which really do still fail in a distributed mode, but they all seem to be in the category "known design decisions wrapped in errorIfDistributedMesh for safety", not anything more urgent.  At first glance it looks like we can probably make half of that code distributed-compatible without tearing up its guts too much, but I suspect it's a low priority, so I'll drop it for now.

**Comment by roystgnr** [comment]
This is now done as far as test/ on gcc is concerned, but I'm leaving the issue open until we have modules/test working as well (it's not even close - 200 tests failing for me on ReplicatedMesh with -p 2) and until we have test/ robust on other compilers (Intel diffs I can blame on icpc's lack of respect for IEEE floating point requirements, but if we're getting clang failures too there may be something to that)

**Comment by brianmoose** [comment]
@roystgnr In case you don't check, the sweep with --distributed-mesh still fails: https://www.moosebuild.org/job/110196/ . It also looks like distributed mesh with recover fails the same way. The straight parallel sweep fails too but in a different way.

**Comment by roystgnr** [comment]
Oh, I've seen those, thanks.  I can replicate both, but fixes won't be soon coming.

With the parallel sweep failure, I understand the problem, it only affects a test object, only triggers in cases where a partition owns no nodes, and is basically an overzealous assertion, so I'm being lazy about figuring out the proper fix for this not-a-real-problem.

With the distributed failure, it's a huge problem, but I don't understand it *at all*.  We're getting like 0.1% error, which is way too high to be a false positive, but way too low for me to track down any obvious root of the problem... and the test case is simple distributed AMR, for which we have hundreds of other test cases inside and outside MOOSE all giving us up-to-solver-tolerance accuracy.

**Comment by permcody** [comment]
This ticket is going to be one of the most referenced, worked on tickets ever... Is it just because of the number? #1500 is pretty easy to remember...

**Comment by roystgnr** [comment]
It's because I'm too lazy to make a new ticket for every --distributed-mesh specific bug I see, unless it's something like #9886 where I don't see how to fix it right away or something like #9869 where the fix is a major undertaking.  Creating a new "issue N: Whatever.C is assuming it's on a ReplicatedMesh" for every "PR N+1: make Whatever.C compatible with a DistributedMesh" feels redundant.  I can start making more specific tickets, though, if you guys need that for stats- or QA-regulation- reasons.  It wouldn't be 100% overhead, even if the issue writing would take as long as the bugfix writing, because generally actually *tracking down* each bug takes 10x as long as either.

**Comment by permcody** [comment]
This is more of an observation. The way you are working now is fine. I'm more concerned about tickets that were opened years ago with no action taken on them ever. Especially if the person who opened them has moved on.

**Comment by YaqiWang** [comment]
You probably should immediately assign the issue to a developer whenever possible. The issue can be reassigned later with a reason.

**Comment by milljm** [comment]
@permcody wow is still an issue with the new scheduler system in place? I haven't seen any hanging parallel mesh tests... since, well I can't remember when.

**Comment by permcody** [comment]
I'm closing this. It's basically done. We'll open new specific issues to address future enhancements.

**Comment by brianmoose** [comment]
Let me just close this for you....
